homeassistant/components/samsungtv/bridge.py

```diff
@@ -409,70 +418,139 @@ def _send_key(self, key: str, key_type: str | None = None) -> None:
         return self._remote
 
-    async def async_close_remote(self) -> None:
-        """Close remote object."""
-        await self.hass.async_add_executor_job(self._close_remote)
+    async def _websocket_event(self, event: str, response: Any) -> None:
+        """Handle websocket event."""
+        LOGGER.debug("SamsungTVWS websocket event: %s", response)
+        if event == "ed.installedApp.get":
+            while self._app_list_futures:
+                self._app_list_futures.pop().set_result(response)
 
-    def _close_remote(self) -> None:
+    async def async_close_remote(self) -> None:
         """Close remote object."""
         try:
             if self._remote is not None:
                 # Close the current remote connection
-                self._remote.close()
+                await self._remote.close()
             self._remote = None
         except OSError:
             LOGGER.debug("Could not establish connection")
+
+
+# This will be moved to the external library
+# It is here to allow testing
+class SamsungTVWSAsyncConnection(BaseSamsungTVWSAsyncConnection):  # type: ignore[misc]
+    """Custom class to extend SamsungTVWSAsyncConnection."""
+
+    connection: WebSocketClientProtocol | None
+
+    def __init__(self, *args: Any, **kwargs: Any) -> None:
+        """Initialise class."""
+        super().__init__(*args, **kwargs)
+        self._recv_loop: Task | None = None
+
+    async def open(self) -> WebSocketClientProtocol:
+        """Open websocket."""
+        if self.connection:
+            # someone else already created a new connection
+            return self.connection
+
+        self.connection = await super().open()
+        return self.connection  # type: ignore[return-value]
+
+    async def start_listening(
+        self, callback: Callable[[str, Any], Awaitable[None]]
+    ) -> None:
+        """Open, and start listening."""
+        if self.connection is None:
+            self.connection = await self.open()
+
+        self._recv_loop = create_task(self._do_start_listening(callback))
+
+    async def _do_start_listening(
+        self, callback: Callable[[str, Any], Awaitable[None]]
+    ) -> None:
+        """Do start listening."""
+        with contextlib.suppress(ConnectionClosed):
+            while data := await self.connection.recv():  # type: ignore[union-attr]
+                response = process_api_response(data)
+                await callback(response.get("event", "*"), response)
+
+    async def close(self) -> None:
+        """Close websocket and stop listening."""
+        if self.connection:
+            await self.connection.close()
+            if self._recv_loop:
+                await self._recv_loop
+            self.connection = None
+
+        await super().close()
+
+    def is_alive(self) -> bool:
+        """Confirm is the connection is available."""
+        if connection := self.connection:
+            return connection.closed
```

Comment:
```suggestion
            return not connection.closed
```

This looks reversed

Reply:
Bad adjustment for mypy.
It is correct in the upstream library: https://github.com/xchwarze/samsung-tv-ws-api/blob/a84318ab878e9453827a1a182e80478057a19365/samsungtvws/async_connection.py#L110-L111